In [5]:
!pip install numpy
!pip install scikit-learn
!pip install pandas


[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: pip install --upgrade pip
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.6/12.6 MB 30.0 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 kB 15.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 505.5/505.5 kB 22.7 MB/s eta 0:00:00

[notice] A new release of pip is available: 23.0.1 -> 24.0
[notice] To update, run: pip install --upgrade pip


# 1. Primeros pasos

El objetivo de este notebook es servir como guía de la implementación del Algoritmo Genético (AG) para el proyecto de la asignatura, así como la revisión de resultados obtenidos y una serie de evaluaciones sobre los distintos hiperparámetros que contiene.

## Inicialización del Algoritmo Genético

El primer paso para trabajar con el AG será la importación de la clase que lo define. Posteriormente, lo inicializamos; para ello requerimos de una serie de atributos, entre ellos una semilla para poder reproducir los resultados aquí obtenidos.

## Trabajo con Datasets

Para la ilustración del algoritmo, trabajaremos con dos datasets, 'toy' y 'housing'.


In [6]:
from AG_class import AG
import os

nombre_dataset = 'toy1'

data_folder = "./data/"

nombre_dataset_train = os.path.join(data_folder, nombre_dataset + "_train.csv")
nombre_dataset_val = os.path.join(data_folder, nombre_dataset + "_val.csv")

ag = AG(
    datos_train=nombre_dataset_train,
    datos_test=nombre_dataset_val,
    seed=123,
    nInd=500,
    maxIter=120
)

               A1          A2          A3          A4          A5           y
count  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000
mean     0.154880    0.064440   -0.021640    0.025360   -0.005470    0.348700
std      0.623641    0.624234    0.542498    0.615866    0.564761    1.626879
min     -0.981000   -0.986000   -0.947000   -0.994000   -0.954000   -3.000000
25%     -0.425500   -0.503500   -0.528500   -0.487750   -0.476500   -0.880000
50%      0.310000    0.161000   -0.053500    0.015000   -0.068500    0.240000
75%      0.676750    0.587250    0.373750    0.626750    0.509750    1.715000
max      0.985000    0.998000    0.990000    0.978000    0.948000    3.930000
              A1         A2         A3         A4         A5          y
count  30.000000  30.000000  30.000000  30.000000  30.000000  30.000000
mean    0.023167   0.196100   0.049667  -0.010100  -0.216533   0.222667
std     0.560692   0.577508   0.539810   0.547763   0.503083   1.463244
min    -0.

nInd representará el tamaño de la población con la que trabajaremos y maxIter el número de generaciones que correrá el algoritmo. Una vez definida la clase, simplemente tenemos que ejecutar la función run() del AG. Haremos uso de el módulo time para poder medir cómo de rápida es la ejecución del AG para el problema dado.

In [7]:
import time

inicio = time.time()
ind, y_pred = ag.run()
fin = time.time()
print(f'Tiempo ejecucion: {(fin-inicio):.2f} segundos')

Generation 0: Best Fitness = 1.0433688300313906
Generation 1: Best Fitness = 1.0433688300313906
Generation 2: Best Fitness = 1.0433688300313906
Generation 3: Best Fitness = 1.0433688300313906
Generation 4: Best Fitness = 0.8853178156899812
Generation 5: Best Fitness = 0.607851482289269
Generation 6: Best Fitness = 0.5478706930348878
Generation 7: Best Fitness = 0.23549489298751725
Generation 8: Best Fitness = 0.23549489298751725
Generation 9: Best Fitness = 0.23549489298751725
Generation 10: Best Fitness = 0.14895926230703135
Generation 11: Best Fitness = 0.14895926230703135
Generation 12: Best Fitness = 0.14895926230703135
Generation 13: Best Fitness = 0.14309998996322026
Generation 14: Best Fitness = 0.09989128185472705
Generation 15: Best Fitness = 0.09859514417187726
Generation 16: Best Fitness = 0.05370357805912208
Generation 17: Best Fitness = 0.05370357805912208
Generation 18: Best Fitness = 0.053376936074340284
Generation 19: Best Fitness = 0.053376936074340284
Generation 20: B

Una vez hemos finalizado, tendremos un mejor individuo obtenido a través de las iteraciones del AG. En nuestro caso, el individuo representa los coeficientes, exponentes y la constante C que han obtenido un mejor fitness para el dataset concreto. Evaluamos ahora su capacidad predictiva con las métricas propuestas

In [8]:
from sklearn.metrics import r2_score,root_mean_squared_error
import pandas as pd

# Imprimir mejor solución encontrada
print(f'Mejor individuo: {ind}')

# Imprimir predicciones sobre el conjunto de test
print(f'Predicciones: {y_pred}')

# Cargar valores reales de 'y' en el conjunto de validacion/test 
# y calcular RMSE y R2 con las predicciones del AG
y_true = pd.read_csv(nombre_dataset_val)['y']  # Cambiar 'target' por 'y'
rmse = root_mean_squared_error(y_true, y_pred)
print(f'RMSE: {rmse:.4f}')

r2 = r2_score(y_true, y_pred)
print(f'R2: {r2:.4f}')

Mejor individuo: [ 1.00097547 -1.01834589  2.01274203 -0.50638633  0.98515626  1.26889671
  1.23715971  1.48627881  1.4469283   1.78800749  0.01396608]
Predicciones: [-1.54399017  1.50997669  2.23289103  0.67146245  0.69799407 -2.55027804
 -0.40382775  2.07723938 -1.17649205  0.32153611  1.72632868 -1.17391997
  1.98692235  2.93147514  1.27151291  1.03164665 -0.68831128 -1.49977283
  1.61643947  1.49547161  1.03273584 -1.55952442 -0.73534336 -1.06740317
  1.14360147 -0.10862479  0.09157355  0.46400599 -0.08236135 -2.82035357]
RMSE: 0.0664
R2: 0.9979


Los valores de RMSE (Root Mean Squared Error) de 0.0664 y un coeficiente R2 de 0.9979 indican un desempeño muy alto del modelo de regresión. El RMSE bajo sugiere que las predicciones del modelo están, en promedio, muy cerca de los valores reales, implicando pocos errores de predicción. Por otro lado, un R2 de 0.9979 muestra que casi toda la variabilidad de la variable dependiente ha sido capturada por el modelo.

# 2. Ajuste de hiperparámetros

En esta sección vamos a trabajar con el dataset 'housing', algo más complejo, que nos permitirá evaluar qué ocurre al modificar ciertos hiperparámetros en cuanto a la velocidad de la convergencia del modelo.

Aplicamos de nuevo el mismo proceso para inicializar el algoritmo.

Nota: Debido al diseño del algoritmo, las variables 'k' y 'mutation_rate' solo se pueden modificar internamente en la clase AG_class.py, por tanto si se quieren obtener los resultados obtenidos a partir de ahora, habrá que realizar dichas variaciones sobre la propia clase.

**K=3,mutation_rate=0.05**

In [9]:
# Nombre generico del dataset
nombre_dataset = 'housing'

data_folder = "./data/"

nombre_dataset_train = os.path.join(data_folder, nombre_dataset + "_train.csv")
nombre_dataset_val = os.path.join(data_folder, nombre_dataset + "_val.csv")

# La clase AG debe estar implementada
ag = AG(
    datos_train=nombre_dataset_train,
    datos_test=nombre_dataset_val,
    seed=123,
    nInd=100,
    maxIter=150
)


             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  11558.000000  11558.000000  11558.000000  11558.000000  11558.000000   
mean       3.853406     28.551479      5.426729      1.097577   1436.749611   
std        1.889429     12.644913      2.463691      0.452346   1182.664306   
min        0.499900      1.000000      1.000000      0.333333      3.000000   
25%        2.555600     18.000000      4.443629      1.005413    788.000000   
50%        3.523250     29.000000      5.223123      1.049339   1168.000000   
75%        4.729300     37.000000      6.028807      1.100291   1736.750000   
max       15.000100     52.000000    141.909091     25.636364  35682.000000   

           AveOccup      Latitude     Longitude             y  
count  11558.000000  11558.000000  11558.000000  11558.000000  
mean       3.156385     35.636775   -119.567083      2.053305  
std       13.675029      2.138824      2.002540      1.151768  
min        0.692308     32.55000

In [10]:
# Ejecucion del AG midiendo el tiempo
inicio = time.time()
ind, y_pred = ag.run()
fin = time.time()
print(f'Tiempo ejecucion: {(fin-inicio):.2f} segundos')

# Imprimir mejor solución encontrada
print(f'Mejor individuo: {ind}')

# Imprimir predicciones sobre el conjunto de test
print(f'Predicciones: {y_pred}')

# Cargar valores reales de 'y' en el conjunto de validacion/test 
# y calcular RMSE y R2 con las predicciones del AG
y_true = pd.read_csv(nombre_dataset_val)['y']  # Cambiar 'target' por 'y'
rmse = root_mean_squared_error(y_true, y_pred)
print(f'RMSE: {rmse:.4f}')

r2 = r2_score(y_true, y_pred)
print(f'R2: {r2:.4f}')


Generation 0: Best Fitness = 1.1200721580177762
Generation 1: Best Fitness = 1.1200721580177762
Generation 2: Best Fitness = 0.760499960893758
Generation 3: Best Fitness = 0.760499960893758
Generation 4: Best Fitness = 0.7481745784693548
Generation 5: Best Fitness = 0.7481745784693548
Generation 6: Best Fitness = 0.7301362245993203
Generation 7: Best Fitness = 0.7301362245993203
Generation 8: Best Fitness = 0.7301362245993203
Generation 9: Best Fitness = 0.6682508619314719
Generation 10: Best Fitness = 0.6682508619314719
Generation 11: Best Fitness = 0.6682508619314719
Generation 12: Best Fitness = 0.6682508619314719
Generation 13: Best Fitness = 0.6682508619314719
Generation 14: Best Fitness = 0.6682508619314719
Generation 15: Best Fitness = 0.6682508619314719
Generation 16: Best Fitness = 0.6662542233938565
Generation 17: Best Fitness = 0.6662542233938565
Generation 18: Best Fitness = 0.6655426779925547
Generation 19: Best Fitness = 0.6655426779925547
Generation 20: Best Fitness = 0.

**K=7,mutation_rate=0.05**

In [11]:
# Ejecucion del AG midiendo el tiempo
inicio = time.time()
ind, y_pred = ag.run()
fin = time.time()
print(f'Tiempo ejecucion: {(fin-inicio):.2f} segundos')

# Imprimir mejor solución encontrada
print(f'Mejor individuo: {ind}')

# Imprimir predicciones sobre el conjunto de test
print(f'Predicciones: {y_pred}')

# Cargar valores reales de 'y' en el conjunto de validacion/test 
# y calcular RMSE y R2 con las predicciones del AG
y_true = pd.read_csv(nombre_dataset_val)['y']  # Cambiar 'target' por 'y'
rmse = root_mean_squared_error(y_true, y_pred)
print(f'RMSE: {rmse:.4f}')

r2 = r2_score(y_true, y_pred)
print(f'R2: {r2:.4f}')

Generation 0: Best Fitness = 1.3547558433973548
Generation 1: Best Fitness = 1.3547558433973548
Generation 2: Best Fitness = 1.3547558433973548
Generation 3: Best Fitness = 1.3143880719230372
Generation 4: Best Fitness = 1.083391255085362
Generation 5: Best Fitness = 1.083391255085362
Generation 6: Best Fitness = 1.083391255085362
Generation 7: Best Fitness = 1.0077064340572257
Generation 8: Best Fitness = 1.0077064340572257
Generation 9: Best Fitness = 1.0077064340572257
Generation 10: Best Fitness = 1.0077064340572257
Generation 11: Best Fitness = 1.0077064340572257
Generation 12: Best Fitness = 1.0077064340572257
Generation 13: Best Fitness = 0.9693985090362326
Generation 14: Best Fitness = 0.9661414906929906
Generation 15: Best Fitness = 0.9393406872585798
Generation 16: Best Fitness = 0.9393406872585798
Generation 17: Best Fitness = 0.9393406872585798
Generation 18: Best Fitness = 0.9373210695695712
Generation 19: Best Fitness = 0.9310183012565743
Generation 20: Best Fitness = 0.9

In [12]:
# Ejecucion del AG midiendo el tiempo
inicio = time.time()
ind, y_pred = ag.run()
fin = time.time()
print(f'Tiempo ejecucion: {(fin-inicio):.2f} segundos')

# Imprimir mejor solución encontrada
print(f'Mejor individuo: {ind}')

# Imprimir predicciones sobre el conjunto de test
print(f'Predicciones: {y_pred}')

# Cargar valores reales de 'y' en el conjunto de validacion/test 
# y calcular RMSE y R2 con las predicciones del AG
y_true = pd.read_csv(nombre_dataset_val)['y']  # Cambiar 'target' por 'y'
rmse = root_mean_squared_error(y_true, y_pred)
print(f'RMSE: {rmse:.4f}')

r2 = r2_score(y_true, y_pred)
print(f'R2: {r2:.4f}')

Generation 0: Best Fitness = 1.0528250807024915
Generation 1: Best Fitness = 1.0528250807024915
Generation 2: Best Fitness = 1.0528250807024915
Generation 3: Best Fitness = 0.8923176551602946
Generation 4: Best Fitness = 0.7550823670064148
Generation 5: Best Fitness = 0.7550823670064148
Generation 6: Best Fitness = 0.7550823670064148
Generation 7: Best Fitness = 0.7550823670064148
Generation 8: Best Fitness = 0.7550823670064148
Generation 9: Best Fitness = 0.7550823670064148
Generation 10: Best Fitness = 0.7006703639400654
Generation 11: Best Fitness = 0.700078446201712
Generation 12: Best Fitness = 0.700078446201712
Generation 13: Best Fitness = 0.700078446201712
Generation 14: Best Fitness = 0.700078446201712
Generation 15: Best Fitness = 0.700078446201712
Generation 16: Best Fitness = 0.700078446201712
Generation 17: Best Fitness = 0.700078446201712
Generation 18: Best Fitness = 0.700078446201712
Generation 19: Best Fitness = 0.700078446201712
Generation 20: Best Fitness = 0.7000784

Como podemos ver, un valor k demasiado grande empeora la convergencia, mientras que un tamaño como k = 5 ha obtenido unos resultados algo mejores. Recordemos que este valor define el tamaño de nuestro torneo, es decir cuantos individuos se seleccionarán al azar, para luego evaluar su fitness y escoger al mejor

Ahora, dejando el valor k = 3 previo, variamos el ratio de mutación.
**K=3,mutation_rate=0.15**